src/main/java/com/nimbusds/jose/crypto/ECDH1PUEncrypter.java

```diff
@@ -0,0 +1,275 @@
+	 */
+	private final SecretKey contentEncryptionKey;
+
+	/**
+	 * Creates a new Elliptic Curve Diffie-Hellman encrypter.
+	 *
+	 * @param privateKey The private EC key. Must not be {@code null}.
+	 *
+	 * @param publicKey The public EC key. Must not be {@code null}.
+	 *
+	 * @throws JOSEException If the elliptic curve is not supported.
+	 */
+	public ECDH1PUEncrypter(final ECPrivateKey privateKey, final ECPublicKey publicKey)
+		throws JOSEException {
+
+		this(privateKey, publicKey, null);
+	}
+
+
+	/**
+	 * Creates a new Elliptic Curve Diffie-Hellman encrypter with an
+	 * optionally specified content encryption key (CEK).
+	 *
+	 * @param privateKey            The private EC key. Must not be
+	 *               			   {@code null}.
+	 *
+	 * @param publicKey            The public EC key. Must not be
+	 *                             {@code null}.
+	 * @param contentEncryptionKey The content encryption key (CEK) to use.
+	 *                             If specified its algorithm must be "AES"
+	 *                             and its length must match the expected
+	 *                             for the JWE encryption method ("enc").
+	 *                             If {@code null} a CEK will be generated
+	 *                             for each JWE.
+	 * @throws JOSEException       If the elliptic curve is not supported.
+	 */
+	public ECDH1PUEncrypter(final ECPrivateKey privateKey, final ECPublicKey publicKey, final SecretKey contentEncryptionKey)
+		throws JOSEException {
+
+		super(Curve.forECParameterSpec(publicKey.getParams()));
+
+		this.privateKey = privateKey;
+		this.publicKey = publicKey;
+
+		if (contentEncryptionKey != null) {
+			if (contentEncryptionKey.getAlgorithm() == null || !contentEncryptionKey.getAlgorithm().equals("AES")) {
+				throw new IllegalArgumentException("The algorithm of the content encryption key (CEK) must be AES");
+			} else {
+				this.contentEncryptionKey = contentEncryptionKey;
+			}
+		} else {
+			this.contentEncryptionKey = null;
+		}
+	}
+
+
+	/**
+	 * Returns the public EC key.
+	 *
+	 * @return The public EC key.
+	 */
+	public ECPublicKey getPublicKey() {
+
+		return publicKey;
+	}
+
+
+	/**
+	 * Returns the private EC key.
+	 *
+	 * @return The private EC key.
+	 */
+	public ECPrivateKey getPrivateKey() {
+
+		return privateKey;
+	}
+
+
+	@Override
+	public Set<Curve> supportedEllipticCurves() {
+
+		return SUPPORTED_ELLIPTIC_CURVES;
+	}
+
+
+	@Override
+	public JWECryptoParts encrypt(final JWEHeader header, final byte[] clearText)
+		throws JOSEException {
+
+		// Generate ephemeral EC key pair on the same curve as the consumer's public key
+		KeyPair ephemeralKeyPair = generateEphemeralKeyPair(publicKey.getParams());
+		ECPublicKey ephemeralPublicKey = (ECPublicKey)ephemeralKeyPair.getPublic();
+		ECPrivateKey ephemeralPrivateKey = (ECPrivateKey)ephemeralKeyPair.getPrivate();
+
+		// Add the ephemeral public EC key to the header
+		JWEHeader updatedHeader = new JWEHeader.Builder(header).
```

Comment:
Do we assume that `skid`, `apu`, `apv` headers are already there? Do we need to check that all these fields are present (probably except skid as it seems to be optional)?

Reply:
They are optional. 
By spec you may don't specify `apu`, `apv` but you will have a bad OtherInfo, but it' allowed.
`skid` is optional as well. Also it's unspecified by spec and I don't see any advantage to trying to implement somehow it. In my opinion DIDComm library SHOULD resolve this header.